src/internal/pachsql/util.go

```diff
@@ -0,0 +1,163 @@
+		default:
+			panic(fmt.Sprintf("driver not supported: %s", driver))
+		}
+	}
+	return schemaName, tableName
+}
+
+// TestRow is the type of a row in the test table
+type TestRow struct {
+	Id int
+
+	Smallint int16
+	Int      int32
+	Bigint   int64
+	Float    float32
+	Varchar  string
+	Time     time.Time
+
+	SmallintNull sql.NullInt16
+	IntNull      sql.NullInt32
+	BigintNull   sql.NullInt64
+	FloatNull    sql.NullFloat64
+	VarcharNull  sql.NullString
+	TimeNull     sql.NullTime
+}
+
+// CreateTestTable creates a test table at name in the database
+func CreateTestTable(db *DB, name string) error {
+	q := fmt.Sprintf(`CREATE TABLE %s (
+		c_id INT PRIMARY KEY NOT NULL,
+
+		c_smallint SMALLINT NOT NULL,
+		c_int INT NOT NULL,
+		c_bigint BIGINT NOT NULL,
+		c_float FLOAT NOT NULL,
+		c_varchar VARCHAR(100) NOT NULL,
+		c_time TIMESTAMP NOT NULL,
+
+		c_smallint_null SMALLINT NULL,
+		c_int_null INT NULL,
+		c_bigint_null BIGINT NULL,
+		c_float_null FLOAT NULL,
+		c_varchar_null VARCHAR(100) NULL,
+		c_time_null TIMESTAMP NULL
+	)`, name)
+	_, err := db.Exec(q)
+	return errors.EnsureStack(err)
+}
+
+func LoadTestData(db *DB, tableName string, n int) error {
+	fz := fuzz.New()
+	fz.Funcs(func(ti *time.Time, co fuzz.Continue) {
+		*ti = time.Now()
+	})
+	fz.Funcs(fuzz.UnicodeRange{First: '!', Last: '~'}.CustomStringFuzzFunc())
+	for i := 0; i < n; i++ {
+		var x TestRow
+		x.Time = time.Now()
+		if i > 0 {
+			fz.Fuzz(&x)
+		}
+		x.Id = i
+		insertStatement := `INSERT INTO test_data ` + formatColumns(x) + ` VALUES ` + formatValues(x, db)
+		if _, err := db.Exec(insertStatement, makeArgs(x)...); err != nil {
+			return errors.EnsureStack(err)
+		}
+	}
+	return nil
+}
+
+func formatColumns(x interface{}) string {
+	var cols []string
+	rty := reflect.TypeOf(x)
+	for i := 0; i < rty.NumField(); i++ {
+		field := rty.Field(i)
+		col := "c_" + toSnakeCase(field.Name)
+		cols = append(cols, col)
+	}
+	return "(" + strings.Join(cols, ", ") + ")"
+}
+
+func formatValues(x interface{}, db *DB) string {
+	var placeholder func(i int) string
+	switch db.DriverName() {
+	case "pgx":
+		placeholder = func(i int) string { return "$" + strconv.Itoa(i+1) }
+	case "mysql", "snowflake":
+		placeholder = func(int) string { return "?" }
+	default:
+		panic(db.DriverName())
+	}
+	var ret string
+	for i := 0; i < reflect.TypeOf(x).NumField(); i++ {
+		if i > 0 {
+			ret += ", "
+		}
+		ret += placeholder(i)
+	}
+	return "(" + ret + ")"
+}
+
+func makeArgs(x interface{}) []interface{} {
+	var vals []interface{}
+	rval := reflect.ValueOf(x)
+	for i := 0; i < rval.NumField(); i++ {
+		v := rval.Field(i).Interface()
+		vals = append(vals, v)
+	}
+	return vals
+}
+
+var matchFirstCap = regexp.MustCompile("(.)([A-Z][a-z]+)")
+var matchAllCap = regexp.MustCompile("([a-z0-9])([A-Z])")
+
+func toSnakeCase(str string) string {
+	snake := matchFirstCap.ReplaceAllString(str, "${1}_${2}")
+	snake = matchAllCap.ReplaceAllString(snake, "${1}_${2}")
+	return strings.ToLower(snake)
+}
```

Comment:
I guess this is doing extra work to deal with overlap issues and lack of lookahead? Since we don't care about numbers or acronyms, I'd kind of prefer something that looks a little more straightforward (hopefully `\B` for "not a boundary/start of line" isn't too obscure)

```suggestion
var matchCaps = regexp.MustCompile("\B([A-Z])")

func toSnakeCase(str string) string {
	snake := matchCaps.ReplaceAllString(str, "_${1}")
	return strings.ToLower(snake)
}
```